### PR TITLE
rgw-lifecycle: add support tag-based filter for versioning bucket

### PR DESCRIPTION
**Pull request by cooboos**
Signed-off-by: Songbo Wang wangsongbo@cloudin.cn

### CODE REVIEW

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -408,29 +436,11 @@ int RGWLC::bucket_lc_process(string& shard_id)
...
-            }
-
-            if (!includes(dest_obj_tags.get_tags().begin(),
-                          dest_obj_tags.get_tags().end(),
-                          prefix_iter->second.obj_tags->get_tags().begin(),
-                          prefix_iter->second.obj_tags->get_tags().end())){
-              ldout(cct, 20) << __func__ << "() skipping obj " << key << " as tags do not match" << dendl;
-              continue;
-            }
+            else if (ret == -EIO)
```
Comment (yehudasa) [inline src/rgw/rgw_lc.cc]: why only `-EIO`? maybe any `ret < 0`?

File: src/rgw/rgw_lc.cc [hunk 2]
```diff
@@ -503,6 +513,16 @@ int RGWLC::bucket_lc_process(string& shard_id)
         bool skip_expiration;
         bool is_expired;
         for (auto obj_iter = objs.begin(); obj_iter != objs.end(); ++obj_iter) {
+          rgw_obj_key key(obj_iter->key);
+          rgw_obj obj(bucket_info.bucket, key);
+          if (prefix_iter->second.obj_tags != boost::none) {
+            ret = check_objtags(cct, store, prefix_iter, bucket_info, obj);
+            if (ret > 0)
+              continue;
+            else if (ret == -EIO)
```
Comment (yehudasa) [inline src/rgw/rgw_lc.cc]: same here

**Comment by mattbenjamin** [comment]
@yehudasa agree, but also, @cooboos I think it would be very helpful to have a few sentences of description in the commit message,and maybe to also create and attach to a tracker ticket?

File: src/rgw/rgw_lc.cc [hunk 3]
```diff
@@ -334,6 +334,34 @@ static int read_obj_tags(RGWRados *store, RGWBucketInfo& bucket_info, rgw_obj& o
   return read_op.get_attr(RGW_ATTR_TAGS, tags_bl);
 }
 
+static inline int check_objtags(CephContext *cct, RGWRados *store, map<string, lc_op>::iterator& prefix_iter, RGWBucketInfo& bucket_info, rgw_obj& obj)
```
Comment (theanalyst) [inline src/rgw/rgw_lc.cc]: consider wrapping this long line into a couple of lines

File: src/rgw/rgw_lc.cc (hunk 1 above)
Comment (dang) [reply src/rgw/rgw_lc.cc]: It only currently returns EIO, but I agree, it should be <0 for future proofing.

**Pull request closed without merging**